src/evict/evict_page.c

```diff
@@ -675,9 +676,40 @@ __evict_review(WT_SESSION_IMPL *session, WT_REF *ref, uint32_t evict_flags, bool
             LF_SET(WT_REC_SCRUB);
     }
 
+    /* Acquire a snapshot if coming through eviction thread route. */
+    if (F_ISSET(session, WT_SESSION_INTERNAL) && !F_ISSET(session->txn, WT_TXN_RUNNING) &&
+      !F_ISSET(conn,
+        WT_CONN_CLOSING | WT_CONN_RECOVERING | WT_CONN_IN_MEMORY | WT_CONN_CLOSING_TIMESTAMP) &&
+      !WT_IS_HS(S2BT(session))) {
+        /*
+         * We don't have a running transaction with eviction threads. Starting a new transaction
+         * should implicitly take a snapshot as well.
+         */
+        WT_RET(__wt_txn_begin(session, txn_cfg));
+        /*
+         * Eviction threads do not need to pin anything in the cache. We have a exclusive lock for
+         * the page being evicted so we are sure that the page will always be there while it is
+         * being processed.
+         */
+        WT_SESSION_TXN_SHARED(session)->pinned_id = WT_TXN_NONE;
```

Comment:
Is this safe? Is it possible that update oldest has already considered the ID (and would that be bad)? Is it then safe to clear it in this way?

Reply:
@agorrod - This is an interesting observation.

- I looked for examples in code where we are managing `pinned_id`. In `__wt_txn_cursor_op()` we update the `pinned_id` without any lock whereas in `__checkpoint_prepare()` we updating while holding global txn lock. I have added a commit 665e667 to acquire a lock before updating `pinned_id`.

- If the `pinned_id` has already been used to calculate global `oldest_id`, there will be a short time window where we can't evict updates belonging to a committed transaction with `pinned_id`. But as `__wt_txn_update_oldest()` is called frequently, we will get out of this condition soon. Also, now that we use snapshots for eviction threads, global visibility checks with `oldest_id` will be a thing of past.